New features

crazoter opened the pull request:
+ Colored bookmarks for entered / owned giveaways (CAUTION: DOESN'T ACCOUNT FOR DLCs?)
+ Marked giveaways as train

Descript:
~~Quick and dirty addition.~~

EDIT 2: Program breaking issues should be fixed. Please check and confirm it is so :) Old bookmarks may appear as entered if the user had entered them previously before this update and vice versa; fixDatabase() doesn't work because Giveaways.loadGiveaway never calls the callback...
Further extensive testing indicates this coloring of bookmarks may be somewhat buggy but generally works as intended. Will continue working on it tmr.
~~EDIT: program breaking issues found in new features, do not merge yet~~
~~+ Upon further testing, the code is currently suffering from race issues; the code if a giveaway has been entered is a hit and miss when you refresh the page; works fine on the page itself, but the bookmark may revert after refreshing. Will debug when I get some free time.  I was interacting with multiple bookmarks via multiple tabs... that must have affected the save code because it's currently using only 1 save file. I'll need to change the save method by ensuring the data is updated first before saving.~~
~~+ Old bookmarks may appear as entered if the user had entered them previously before this update and vice versa.~~
~~I've not implemented a database/giveaway check etc. Should be quite straight forward.~~

Possible Issues (No DLC detection):
The only possible issue is that I only used the "isOwned" variable to check if a giveaway is owned and did not account for DLCs before marking the bookmark in red. Minor issue, since very few people, if any, are going to bookmark games that they cannot enter.

Implementation:
I used the descriptions of the bookmarks to identify if they are part of train. Simply put, if their descriptions are the same (by the same user and generally ending in the same time), I assume the giveaways are together. Works pretty good. For the entered/owned giveaways, I simply leveraged on the sgapi and put a click handler on the enter/exit giveaway button. Clicking will automatically trigger the buildNavRows and saving of bookmark data to reflect changes. 

Preview:
![Image](http://i.imgur.com/KbMpiRo.png)

Future optimizations: This was a *really* quick fix so there's some code bloat that shouldn't affect users. The added styles can be edited into the css files and the addBookmarkMenuItem function header can be optimized to remove the "hasEnded" variable amongst other nitpick optimizations. It otherwise works as intended.

EDIT: The ellipsis now cover only the title, leaving the points still visible if a train giveaway bookmark's title gets too long

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,36 +17,67 @@
 /*jshint multistr: true */
 
+//GA Entered classes
+GM_addStyle(".nav__relative-dropdown.___mh_bookmark_outer_container {z-index:1000;}");
```
Please move the CSS declarations to the *.css file

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,36 +17,67 @@
 use(SgApi);
 use(Util);
 
+//GM_setValue("fixedEntered",false);
```
Please remove all commented out lines

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,36 +17,67 @@
 
 function fixDatabase(){
-	if(parseBool(GM_getValue("fixedDb0.7", false))){
+	var fixedEntered = true;//GM_getValue("fixedEntered",false);
```
You need to parse the result from GM_getValue using parseBool()

crazoter commented:
Okay, I'll look into it later this evening. Thanks for the pointers :)

crazoter commented:
1. Fixed lol. I don't know if the problem is solved but save() -> saveData().
2. I thought about this and here are my thoughts:

Color: I agree that it is too colorful and changed it to use opacity instead. However, I disagree with icons and badges; going by that train of thought, steam gifts should have added an icon/badge to indicate entered giveaways and you should have done the same for expired giveaways. The reason it was not so is because color/opacity is a much clearer indicator than icons/badges. I'll change the color change to an opacity change instead to emulate the way steam gifts indicates entered giveaways but for owned giveaways, I'll retain the red color (since it's a rare use case scenario so it's not really a priority). Regarding :hover, I think I'll change the color to make it nicer.

Indentation: This one is trickier to beautify but I agree that the indentation is over-the-top. I reduced the indentation and added a inset shadow to train giveaways. I have tried the use of icons and badges but found that using them would cluster the giveaways and make it uglier and harder to read.

Changes look like:
![image](https://cloud.githubusercontent.com/assets/9276766/24504085/70a01fc2-1587-11e7-8a8a-7a1ed1ebff77.png)

3. That's kind of a low priority at the moment. I'll do it when I have time.
4. I'll keep it in GM_addStyle until the CSS is confirmed, since the CSS file is linked to the github repo.

Edit: Whoops, you replied to my old post :P. For archiving purposes, I had originally posted an image similar to the one maherm posted below but using the train icon from Font Awesome pack which Steam Gifts currently uses.

maherm commented:
Still not sold on the colors, Sure, it is a clearer indicator for the state, but only because it is much more invasive than icons, which is the part i dislike about this solution. It makes the bookmarks list look pretty inconsistent. 

I like that new indentation!
You're rigtht about the train symbol, it does look like a skull :-D You could try u+1F682 (Steam Locomotive) instead, or what i would like the most, try and mimic a train map, like this (badly painted) example:
![7e618b60-157c-11e7-8eb0-5b91fa2cd06d](https://cloud.githubusercontent.com/assets/5812806/24504022/da022aba-1554-11e7-83ff-c16ab99f9155.png)
If you need additional icons for that, please embedd them as svg path or base64 image.

crazoter commented:
I like that train map idea. Will look into.
I'm pretty satisfied with how it turned out. Is it better than the simpler indent + shadow? Or maybe it needs more opacity?
![image](https://cloud.githubusercontent.com/assets/9276766/24507245/8c883ff2-1592-11e7-9eed-f3fc8d29b2a4.png)

More example
![image](https://cloud.githubusercontent.com/assets/9276766/24507406/eeffe072-1592-11e7-9987-e86130194a2f.png)

I tried adding a train at the start of the tracks, but I don't think it will work out very well since the image will interfere with it. It'll be *really* cool, but it's not worth the effort :p
![image](https://cloud.githubusercontent.com/assets/9276766/24508677/c4c5a3ce-1596-11e7-8c35-b98e1921afc0.png)

maherm commented on sg_bookmarks.user.js:
```diff
@@ -83,7 +88,7 @@ function fixDatabase(){
 				delete ga.descriptionHtml;
 				data.bookmarks[k] = unwrap(ga);
-				save();
+				saveData();
```
Thanks for fixing that :-)

maherm commented on sg_bookmarks.user.js:
```diff
@@ -24,18 +24,23 @@
+GM_addStyle(".nav__row.__mh_bookmark_item.mid_train{box-shadow: inset 0 0 10px -9px #000000;height: 20px;overflow:hidden;margin: 0px 0px 0px 2px;}");
 GM_addStyle(".nav__row.__mh_bookmark_item.mid_train .nav__row__summary__name{white-space: nowrap;}");
 GM_addStyle(".nav__row.__mh_bookmark_item.mid_train .nav__row__summary__description{display:none;}");
+GM_addStyle(".nav__row.__mh_bookmark_item.mid_train .train_tracks {opacity: 0.3; padding: 0px 4px 0px 4px}");
```
I'd like to decrease the opacity a little bit more. Padding changes are for adjusting the padding from my first comment.
{opacity: 0.2; padding: 0px 6px 0px 3px}

maherm commented on sg_bookmarks.user.js:
```diff
@@ -24,18 +24,23 @@
-GM_addStyle(".nav__row.__mh_bookmark_item.mid_train{height: 20px;overflow:hidden;margin: 1px 1px 1px 10px;}");
+//GM_addStyle(".nav__row.__mh_bookmark_item{border-top:1px solid #DDD;}");
+//GM_addStyle(".nav__row.__mh_bookmark_item{z-index:9;position:relative;box-shadow: 0px 1px 5px -1px #ccc;}");
+GM_addStyle(".nav__row.__mh_bookmark_item.mid_train{box-shadow: inset 0 0 10px -9px #000000;height: 20px;overflow:hidden;margin: 0px 0px 0px 2px;}");
```
Please lose the margin. Use padding-left: 2px isntead

crazoter replied:
Yes I like this too! Looks very neat :)

maherm commented on sg_bookmarks.user.js:
```diff
@@ -24,18 +24,23 @@
-GM_addStyle(".nav__row.__mh_bookmark_item.mid_train{height: 20px;overflow:hidden;margin: 1px 1px 1px 10px;}");
+//GM_addStyle(".nav__row.__mh_bookmark_item{border-top:1px solid #DDD;}");
+//GM_addStyle(".nav__row.__mh_bookmark_item{z-index:9;position:relative;box-shadow: 0px 1px 5px -1px #ccc;}");
+GM_addStyle(".nav__row.__mh_bookmark_item.mid_train{box-shadow: inset 0 0 10px -9px #000000;height: 20px;overflow:hidden;margin: 0px 0px 0px 2px;}");
```
Please prepend all css classes with "___mh_" for namespacing

maherm commented on sg_bookmarks.user.js:
```diff
@@ -24,18 +24,23 @@
-GM_addStyle(".nav__row.__mh_bookmark_item.__mh_state_entered {background-image: linear-gradient(#f6f6ec 0%, #e9e9ca 100%);\
-background-image: -moz-linear-gradient(#f6f6ec 0%, #e9e9ca 100%);\
-background-image: -webkit-linear-gradient(#f6f6ec 0%, #e9e9ca 100%);}");
+GM_addStyle(".nav__row.__mh_bookmark_item.__mh_state_entered *{opacity:0.7;}");
```
I like this a lot more than the yellow background. Looks neat.

maherm commented on sg_bookmarks.user.js:
```diff
@@ -3,7 +3,7 @@
 // @namespace    http://steamgifts.com/
 // @version      0.9
```
Don't forget to increase the version number.

crazoter commented:
I'm going to improve on the icons and add the settings. Any idea how we should add the settings? I'm thinking just put a header row with a gear icon that pops up a modal that has the two options.

maherm commented:
Sorry, I should have been more specific about that.
SgApi offers an interface to generate a Settings Options in the SG Settings. Check the initSettings() method and the [docs over at SgApi ](https://github.com/maherm/sgapi/blob/master/doc/SETTINGS.md).

Basicly, you just need to add 
```javascript
.bool("Visualize trains", true)
.bool("Highlight entered and owned giveaways", true)
```
to the initSettings function and check if the setting is on right before you apply the feature
```javascript
if(data.settings.get("Visualize trains")){
  //do your stuff
}
```

crazoter commented:
![woah](https://image.slidesharecdn.com/pwningwithpowershell-150725153308-lva1-app6892/95/pwning-with-powershell-22-638.jpg?cb=1437838947)

That's *way* easier than I expected it to be

crazoter commented:
![image](https://cloud.githubusercontent.com/assets/9276766/24552975/277abb8e-165b-11e7-89e1-45c8b38462b2.png)
Was trying something out, but the train thing looks worse than I hoped.

![image](https://cloud.githubusercontent.com/assets/9276766/24553144/bbba25aa-165b-11e7-8bd5-b6c6de68e76f.png)
The opacity change helps but... feels out of place.

maherm commented:
> The opacity change helps but... feels out of place.

I think so, too.
I like the new image for the last stop. Why don't you try to put the train on this one?

crazoter commented:
Train in front experiments:
Fiddled around (font awesome icon)
![image](https://cloud.githubusercontent.com/assets/9276766/24553667/7b045cfe-165d-11e7-8727-083759248fc9.png)
*ugh*

Using old sprite
![image](https://cloud.githubusercontent.com/assets/9276766/24553844/2eed941a-165e-11e7-8299-f628bf9865f5.png)
*eurgh*

I'll try the train at last stop idea. 
Prototype
![image](https://cloud.githubusercontent.com/assets/9276766/24554059/d9796c6a-165e-11e7-8600-145141b1ddfe.png)
*this is actually pretty good. May look clunky for 2-length trains though*

Hmm, should we go for the train @ last stop, or just cut it off like:
![image](https://cloud.githubusercontent.com/assets/9276766/24554186/4635d104-165f-11e7-9635-5c8f3c878716.png)
*this isn't too bad either*

Either way, the train at first stop's an experiment gone bad. Gonna scrap that idea.

Gonna wrap up for the day and continue tmr. I think the train at the end idea is the way to go. 👍 
TODO: Fix that bug that sometimes erroneously displays entered GAs as unentered and vice-versa. I'm going to assume it is because I clicked enter GA then bookmarked it before the enter registered.

maherm commented:
I'm fine with both, train at the end or no train at all.

crazoter commented:
![image](https://cloud.githubusercontent.com/assets/9276766/24574493/4b2a4b24-16c6-11e7-870a-a611d3e382bb.png)
*cool and good*

Looks good. I'll optimize then fix problems and wrap this up.

EDIT 1: Problems fixed, the user can go crazy using multiple tabs / super fast navigation and it won't mess up the bookmark states.

EDIT 2: regarding the settings, they are inside and working but I'm unsure if it's currently working as intended; might want to check and confirm.

EDIT 3: "fixDatabase()" doesn't work as intended; it wipes out your bookmarks first and instead of appending each bookmark after every AJAX call, it clears the bookmarks so that you end up with only 1 bookmark even after I waited for all the AJAX calls. This is a dangerous function to debug so I implemented a primitive import / export function for debugging purposes. I had intended to use it so that previous giveaways will also have their entered/owned state indicated but seeing as it is now, I'm going to stop short of doing that until you can ID the issue.

EDIT 4: There is this *annoyingly* persistent bug of finding GAs that are shown as not entered but are entered. I need to investigate this.

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,28 +17,59 @@
 // @grant        GM_listValues
 // @grant        GM_deleteValue
 // @grant        GM_getResourceText
+// @grant        GM_addStyle
```
Wen won't need that one anymore

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,28 +17,59 @@
+//Settings
+var SETTINGS_STATES = "Show giveaway status";
+var SETTINGS_TRAIN = "Group giveaways in a train";
+var settings = new Settings("SG_Bookmarks")
```
You shouldn't create a new Settings-object. Every Settings-object gets it's own section in /account/profile/. We want your Settings to be in the already existing "SG Bookmarks" sections. So you need extend the existing one (data.settings) in the method initSettings (Line 342) and then use that one.
Also, don't fiddle around too much with the settings menu. I'm sitting on a merge for the SgApi Settings Module that brings some new features and bugfixes (like description not being shown for some settings)

maherm commented on sg_bookmarks.user.js:
```diff
@@ -17,28 +17,59 @@
+var SETTINGS_TRAIN = "Group giveaways in a train";
+var settings = new Settings("SG_Bookmarks")
+		.boolean("Show giveaway status", true, {description:"Dim entered giveaways and color giveaways you cannot enter.", values: ["No", "Yes"]})
+    .boolean("Group giveaways in a train", true, {description:"Show which giveaways are in the same train.", values: ["No", "Yes"]});
```
You can lose the ["No","Yes"] array. That's the default value. You only need to pass the ```values``` option if you want the options to be labeled differently.

crazoter commented:
I'm currently still trying out the sync feature. The purpose of the sync feature was to ensure the state of giveaways (i.e. entered or not) are properly recorded, but it slows the process down and is not 100% failproof (entering GAs through another device will wreck it) so I'm thinking about removing it in favour for alternatives; right now I'm considering using the https://www.steamgifts.com/giveaways/entered page for that purpose.

To explain the syncing system, I'll start with the issue it was meant to solve. The current problem is that when you add/update a bookmark the bookmark could have been outdated for 2 reasons: 
1. it was entered on another page tab as such rendering the giveaway on the page outdated or
2. it was bookmarked while the system was in the midst processing the enter giveaway request; a race hazard, in a sense.

In order to catch these cases, I used AJAX (your Giveaway.loadGiveaway) to retrieve the giveaway's most current state whenever I bookmark something. This introduces new problems because AJAX can be interrupted by navigating to other pages, so I had to implement a "buffer" (queuedBookmarkIds) in which giveaway IDs are stored via GM_setValue; this buffer would be persistent across tabs and would ensure the AJAX response is received (by resending the request if it was not completed even though the user is on another page), thereby ensuring the integrity of the state of the giveaway.

As much as I want to say this was the ultimate solution to this issue, this didn't completely solve the problem because it would slow down the process of adding bookmarks (since now there's AJAX involved) and would *still* sometimes register entered giveaways as unentered (for reasons unknown that I've yet to investigate) and giveaways entered via other devices wouldn't have been updated to reflect such changes.

The new solution I'm thinking of would be to use AJAX to scrape all on-going entered giveaways from https://www.steamgifts.com/giveaways/entered page (and page 2 3 etc simultaneously), merge the paginated results into a javascript object with the key being the giveaway ID and use this object to check every bookmark if they are entered. This would be the ultimate solution but it would (probably) be quite intensive if done on every page load since in practice you would be also using AJAX to load around 3 other pages at the same time, so I'm currently still thinking about the when and how I should execute this.

`// @grant        GM_addStyle`
Good catch.

> You shouldn't create a new Settings-object. Every Settings-object gets it's own section in /account/profile/. We want your Settings to be in the already existing "SG Bookmarks" sections. So you need extend the existing one (data.settings) in the method initSettings (Line 342) and then use that one.
Also, don't fiddle around too much with the settings menu. I'm sitting on a merge for the SgApi Settings Module that brings some new features and bugfixes (like description not being shown for some settings)

Ah! I had completely missed the existing settings lol. I will fix in the next commit.

maherm commented:
> As much as I want to say this was the ultimate solution to this issue, this didn't completely solve the problem because it would slow down the process of adding bookmarks (since now there's AJAX involved) 

Why would it slow down the adding? I'd assume the adding is done instantly and the AJAX call is only done to update the state, which would be perfectly acceptable.

> The new solution I'm thinking of would be to use AJAX to scrape all on-going entered giveaways from https://www.steamgifts.com/giveaways/entered page (and page 2 3 etc simultaneously), merge the paginated results into a javascript object with the key being the giveaway ID and use this object to check every bookmark if they are entered. This would be the ultimate solution but it would (probably) be quite intensive if done on every page load since in practice you would be also using AJAX to load around 3 other pages at the same time, so I'm currently still thinking about the when and how I should execute this.

I don't think this is the way to go, because it yields a bunch of other problems: 
- How many pages to fetch? All of them until all bookmarks are updated? All of them until the ended GAs start? What if we don't have any bookmarks, do we still need to fetch 5+ pages? 
- I want the load for the SG servers to be as little as possible. On every page load is not acceptable. A scheduled refresh paired with a buffer like your initial idea might be okay. 

> and giveaways entered via other devices wouldn't have been updated to reflect such changes.

Actually, I initially planned a sync feature using Dropbox, but sadly [they shut down the API I wanted to use](https://www.steamgifts.com/discussion/kceto/userscript-sg-bookmarks-save-giveaways-for-later-use#nyQNSZ2), which gave you an automatically synced key-value-store. The new API only lets you synchronize files. Also, every user needs to create an own dev api key to use it. All of this made me postpone this feature to never-ever-land.

But, if you are already thinking about syncing it may be interesting to consider implementing syncing via Dropbox. It could imitate a key-value store like GM_getValue and we could integrate it into SgApi, which would be pretty awesome.

crazoter commented:
> Why would it slow down the adding? I'd assume the adding is done instantly and the AJAX call is only done to update the state, which would be perfectly acceptable.

I thought that would be the case, but it is actually quite a bit slower to add bookmarks after I implemented syncing. This may be due to one of a few reasons:
1. AJAX async may be set to false for Giveaway.loadGiveaway but that's probably not the case I hope.
2. My browser is lagging because I didn't restart it for a few days already
3. I have too many bookmarks (my last bookmark data export was about 90615 characters long, I can imagine it's about in the 100k++ by now even though I always clean out the expired giveaways)

Well, I'll look into it and keep you in the loop. Hmm, speaking of too much data... Maybe we should remove some attributes like steamUrl, steamAppId, contributorLevel, creator, isWhitelist, isGroup, commentCount for storage purposes.

> they shut down the API I wanted to use
> it may be interesting to consider implementing syncing via Dropbox.

Well, Dropbox's API is gone so we'd have to seek out alternatives. Frankly, I think going from a purely client-side approach is good enough, because chances are the other devices he uses won't be able to execute userscripts and it's not really value-for-effort to implement multi-device syncing.

If you really want multi-device syncing, there's 2 approaches:

1. Authenticate on our own and then store on our own server sort of method
You'd want to look into options like Backend-as-a-Service (e.g. [Backendless](https://backendless.com/pricing/backendless-cloud-pricing/)).
2. Use a third party to authenticate and then store on said third party's storage
I faced this dilemma before and I used Imgur API; what I did was upload JSON as the description and used the image ID as a way to reference that data. I don't intend to do that (ever again), but such options are available. 😆 
*now that I mention it, it was surprisingly effective.*
Jokes aside, you can look into [Reddit API](https://www.reddit.com/dev/api/).

Now you may be wondering how approach 2 is feasible. The logic is this: let's say you use Reddit API. You ask the user to login and then use the API to access a private subreddit that contains all of his bookmarks that you have uploaded using the API. How the bookmarks are contained is entirely up to you; it can even be stuffed into a single comment but the idea is the same.

>  A scheduled refresh paired with a buffer like your initial idea might be okay.

Yup, this is what I'm thinking about.

maherm commented:
> AJAX async may be set to false for Giveaway.loadGiveaway but that's probably not the case I hope.

It's not.

> I have too many bookmarks (my last bookmark data export was about 90615 characters long, I can imagine it's about in the 100k++ by now even though I always clean out the expired giveaways)
Well, I'll look into it and keep you in the loop. Hmm, speaking of too much data... Maybe we should remove some attributes like steamUrl, steamAppId, contributorLevel, creator, isWhitelist, isGroup, commentCount for storage purposes.

Tbh I never thought this would be a problem. Reducing storage data is an option, but we should be careful about what we remove, because if somebody goes ahead and implements a nice method to display level/whitelist/group status/whatever, we'll face the problem that we need a new (hopefully this time working) fixDatabase(). And I'd like to avoid that.
How many bookmarks do you have? Do you really use that much or is it just for developing?
Btw, I already implemented a feature to auto-remove expired GAs, but I havent pushed it yet because I was pretty busy (and the SG thread has been pretty silent for a while now, which doesn't really help to keep the motivation up ;-) )

> Well, Dropbox's API is gone so we'd have to seek out alternatives. 

There is still is the file-based API, which would be a little bit annoying to use, but is imo still the best option.

> Authenticate on our own and then store on our own server sort of method
You'd want to look into options like Backend-as-a-Service (e.g. Backendless).

Nah. Like I wrote in the SG thread, I want users to be in full control of their data. Also, I want a solution that can be integrated in SgApi, and I don't want a single centralized backend for that.

> Use a third party to authenticate and then store on said third party's storage
I faced this dilemma before and I used Imgur API; what I did was upload JSON as the description and used the image ID as a way to reference that data. I don't intend to do that (ever again), but such options are available. 😆
now that I mention it, it was surprisingly effective.
Jokes aside, you can look into Reddit API.

I like your out-of-the-box thinking :-D Still think a file-based solution using Dropbox (or maybe Google Drive) is the better solution.

crazoter commented:
> How many bookmarks do you have? Do you really use that much or is it just for developing?

I bookmark the games that I find from trains and puzzles (that I want to join), sometimes good public giveaways that I don't want to miss. That's like... Uh...
![image](https://cloud.githubusercontent.com/assets/9276766/24635277/16db0cc8-1905-11e7-9991-5e400a23018f.png)
*that's like half of it, the other half can't be shown because firefox could only zoom out to 30%*

lol

> I already implemented a feature to auto-remove expired GAs

Nice. It was kind of annoying at first since it was kind of laggy (since I have so many bookmarks) but I quickly learnt that double clicking removed 2 GAs at once, so it wasn't so bad.

> but we should be careful about what we remove

You're right, I had not considered that.

I think we can shorten the name of the attributes to maybe 1 or 2 characters and use getters/setters instead. It would certainly reduce a lot of space. This change would require some sort of data conversion method for existing users but the storage space saved is pretty decent. Not worth it if you're just loading it from the GM database, but if you're shifting it to storage online it's worth looking into.

example of saved giveaway:
`{"id":"xxxxx","gameTitle":"Frozen Synapse Prime","cp":25,"steamUrl":"http://store.steampowered.com/app/328600/","steamAppId":"328600","copies":1,"contributorLevel":0,"url":"https://www.steamgifts.com/giveaway/2Jy5x/","creator":"adminst","startTime":"1490965948","endTime":"1491030660","isWhitelist":false,"isGroup":false,"hasStarted":true,"hasEnded":false,"isRunning":true,"commentCount":0,"entryCount":48,"thumbUrl":"https://steamcdn-a.akamaihd.net/steam/apps/328600/capsule_184x69.jpg","isOwned":false,"isEntered":true}},"settings":{}}`
*537 char (537B)*

that can be shortened to like
`{"id":"xxxxx","a":"Frozen Synapse Prime","b":25,"c":"http://store.steampowered.com/app/328600/","d":"328600","e":1,"f":0,"url":"https://www.steamgifts.com/giveaway/2Jy5x/","g":"adminst","h":"1490965948","i":"1491030660","j":false,"k":false,"l":true,"m":false,"n":true,"p":0,"q":48,"r":"https://steamcdn-a.akamaihd.net/steam/apps/328600/capsule_184x69.jpg","s":false,"t":true}},"u":{}}`

steamID is pretty much a derived variable from steamAppUrl so we can remove that too (we don't remove steamAppUrl since it can be "/sub/id" instead of "/app/id". However, we *can* shorten the steam app url itself by removing store.steampowered.com. Steam gifts Url is pretty much going to be https://www.steamgifts.com/giveaway/ + id all the time so we can remove that too; it's simply derived from id. The true/false can all be converted to 0 and 1, shortening it even further:

`{"id":"xxxxx","a":"Frozen Synapse Prime","b":25,"c":"app/328600/","e":1,"f":0,"g":"adminst","h":"1490965948","i":"1491030660","j":0,"k":0,"l":1,"m":0,"n":1,"p":0,"q":48,"r":"https://steamcdn-a.akamaihd.net/steam/apps/328600/capsule_184x69.jpg","s":0,"t":1}},"u":{}}`
*265 char (265B)*
That's almost 50% less space required per giveaway if we implement the optimizations.

crazoter commented:
I thought about it a little at work today and I figure the reason why it's lagging is mostly because of the buildNavRows and how it's building so many bookmarks everytime anything changes. I'm going to improve it by ~attaching IDs to the DOMs and~(apparently it's already in) altering them directly instead. It should get rid of the lag and whatsoever. Reducing the storage footprint is definitely a plus if we intend to store it somewhere apart from the PC, in which case we can use the optimizations as outlined above. However, I'll prioritize the UI optimizations and sync fix.

maherm commented:
> {"id":"xxxxx","a":"Frozen Synapse Prime","b":25,"c":"app/328600/","e":1,"f":0,"g":"adminst","h":"1490965948","i":"1491030660","j":0,"k":0,"l":1,"m":0,"n":1,"p":0,"q":48,"r":"https://steamcdn-a.akamaihd.net/steam/apps/328600/capsule_184x69.jpg","s":0,"t":1}},"u":{}}
265 char (265B)

So we are basicly using dictionary based compression with a predefined dictionary. 
1. We can lose the thumbUrl, it is created from the appId or subId, so we can reconstruct it from "c"
2. Were saving a string, and we already need to do some replacements, so why don't we take it to the max? ;-)

```javascript
var a = '{"id":"xxxxx","a":"Frozen Synapse Prime","b":25,"c":"app/328600/","e":1,"f":0,"g":"adminst","h":"1490965948","i":"1491030660","j":0,"k":0,"l":1,"m":0,"n":1,"p":0,"q":48,"s":0,"t":1}},"u":{}}';
console.log(a.length); //190
var compressed = a.replace(/ä/g, "\\ä").replace(/ö/g,"\\ö").replace(/ü/g,"\\ü").replace(/`/g,"\\`"); //Escape whatever we use as replacements
compressed = compressed.replace(/","/g,"ä").replace(/":"/g,"ö").replace(/":/g,"ü").replace(/,"/g,"`"); //Replace
console.log(compressed); //{"idöxxxxxäaöFrozen Synapse Primeäbü25`cöapp/328600/äeü1`fü0`göadminstähö1490965948äiö1491030660äjü0`kü0`lü1`mü0`nü1`pü0`qü48`sü0`tü1}}`uü{}}
var decompressed = compressed.replace(/ä/g,'","').replace(/ö/g,'":"').replace(/ü/g,'":').replace(/`/g,',"'); //Decompress
//... TODO: replace the escaped chars back
console.log(compressed.length) //141
console.log(decompressed == a); //true
```

crazoter commented:
Wow, the syncing feature by entered GA page exceeded my expectations! I synced all my bookmarks so fast I couldn't even believe it. Will include the implementation inside the next commit but won't bring it into use yet.

crazoter commented:
I tried moving settings over to data.settings but it didn't work out. Tsk. I'll work on other stuff before tackling that.

maherm commented:
Did you add your settings before the init() call? Every attempt to add settings ater init() was called will fail.

crazoter commented:
Yup. I added them before the existing bool. I don't know why, but when I use the exportData method the settings would always be shown as empty ("{}"). Will look into.

crazoter commented:
Yup. I added them before the existing bool. I don't know why, but when I use the exportData method the settings would always be shown as empty ("{}").

maherm commented:
That is intended. The Settings object is just an accessor that exposes the functions get(), set() and deleteAll() (which resets all settings to their default values). You are not supposed to directly access the data.

crazoter commented:
> That is intended
Okay, I thought it was a bug for a moment :)

I managed to get it working :)
Btw, should we use useGmStorage for the settings? What implication does this have? Will this affect the use of the app for users who always use private browsing?

EDIT: With the new remove fix, removing giveaways is now really quick and easy even when you have a ton of giveaways. There's no need for the auto remove expired giveaways imo.

maherm commented:
> Btw, should we use useGmStorage for the settings? What implication does this have? 

GM_Storage is always script-private and propably consistent even in private mode (never tested it).
I chose localStorage as default option in SgApi because you don't need any @grants for it.

> Will this affect the use of the app for users who always use private browsing?

Never really thought of that. It certainly does, because by default localStorage is used. 
But I'm not sure whether we should consider this. The vast majority of SG's users (and internet users overall) do not use the private mode for everyday surfing. Switching from localStorage (current) to GM_Storage means people will either lose their settings (not acceptable) or we need to provide a migrate-function, and I'm not willing to put in the extra work for that considering it would affect only "that one guy". If you're up for it, go ahead :-)

crazoter commented:
https://www.elie.net/blog/privacy/19-of-users-use-their-browser-private-mode#.VGHmRodAr5k

crazoter commented:
Whoops, wrong button :p
Fixed the sync all feature, now it really works. The previous version had like 5 bugs in it lol.

crazoter commented:
:/ I think the entered-but-not-shown-as-entered GA bugs may be due to old corrupted data that I created while I was trying to add the feature. I can't seem to replicate it.

*Awesome*

EDIT: ~Wait a minute, what if I enter a giveaway before the userscript even loaded?~ Never mind, that's too hard to execute.

TODO LIST:
1. ~Execute "sync all" on first update, afterwhich after 1hr of inactivity by the user~ DONE, but may want to change 1hr to some other duration
~By "inactivity" i'm thinking along the lines of "user hasn't accessed SG via this computer for the last 1hr". This can be executed with a simple GM value timestamp. Whether or not 1hr is too long is... well, still very much experimental.~
2. ~Optimize adding of bookmark.~ Done
~Right now, it can take about 1-2s to add a bookmark upon entering a GA page. I'm going to look into how I can make this a lot faster; should have a lot to do with the buildNavRows.~
3. ~Execute BuildNavRow only after bookmark icon is clicked because it's currently lagging steamgifts up everytime i navigate?~ Done
4. ~Not something that needs to be implemented but I'm still not sure how to config the settings as a user.~
5. ~...Syncing owned games is currently the most challenging issue. What I'm thinking is every time the user syncs with Steam, sync all is executed. Sync all then checks for bookmarks that were originally entered, but now is no longer entered. In such events, these bookmarks would be synced individually.~ Sync all should manage owned games now, clicking sync with steam will also sync bookmarks. This doesn't account for other userscripts that sync with steam, but it's still covered by the hourly sync.
6. ~Clean up code for deployment~
7. Fixed random bug I didn't know existed!

All done! The only issues now is maybe tweaking the 1hr but apart from that, it's good to go. I'm pretty satisfied with what's in there now.

maherm commented:
Nice!

1. How do you define inactivity? 
2. ??
3. Isn't that how it is now?
4. https://www.steamgifts.com/account/settings/giveaways#SGBookmarks

crazoter commented:
Interesting, I don't see it in the sidebar... Maybe because I'm using another script or something.
EDIT: Nope, still don't see it after removing the userscript. Maybe because I'm using private browsing; in which case, nvm.
![image](https://cloud.githubusercontent.com/assets/9276766/24730120/58f75ea0-1a94-11e7-9760-ce5abcbb00b5.png)

maherm commented:
> Interesting, I don't see it in the sidebar... Maybe because I'm using another script or something.
EDIT: Nope, still don't see it after removing the userscript. 

1. The sidebar in your screenshot is not vanilla SG. Those filters are from a third party app. Are you sure you disabled all other scripts/add-ons?
2. That's the main page sidebar. The SG Bookmarks entry is in the sidebar on the SG Settings page: https://www.steamgifts.com/account/profile/sync
3. Out of curiosity: which browser are you using?

>  Maybe because I'm using private browsing; in which case, nvm.

So you are “that one guy“ that's surfing in private mode? lol sorry :-D
That should have no effect on the page though. It only means your settings are back to default after you close your browser.

crazoter commented:
> That's the main page sidebar. The SG Bookmarks entry is in the sidebar on the SG Settings page: https://www.steamgifts.com/account/profile/sync

Ohhhhhhhh so that's where it is! I found it :p
I'm using Firefox :)

I'm using other userscripts like SG Game Tags, SG Group Info hover etc that aren't too disruptive with the UI. The one you've seen earlier is Bettersteamgiftfilters.

crazoter commented:
I'm currently not in a fresh state of mind so the code is not thoroughly tested yet. Seems to be working well, though!
EDIT: Tested it somewhat. Pretty solid.
[TODO](https://github.com/maherm/steamgifts_scripts/pull/3#issuecomment-291911112) UPDATES
2. Optimize adding of bookmark.
I'm lazy to rework the buildNavRows to become iterative (i.e. insertion sort) instead of the destroy-then-rebuild approach that is currently used now. As such, I added a new var "rebuildNavRows". Simply put, if this variable is set to true, the app will buildNavRows() when the openBookmarkContainer is executed (instead of executing it whenever there's an update which causes massive lag as the browser is rebuilding the DOM unnecessarily), after which rebuildNavRows is set back to false. This also applies to any existing code that requires running buildNavRows() e.g. saving bookmarks. 
3. This is related to 2. I removed the buildNavRows in addNavButton(), now buildNavRows is only executed when the user clicks the button (refer to rebuildNavRows above), so navigating pages in SG doesn't cause the lag anymore for people with massive bookmark folders like me.

TLDR: Navigating pages in SG and adding new bookmarks is now lightning fast (on top of entering/exiting GAs and removing bookmarks, which are already optimized) even if you have a *crazy* number of bookmarks

crazoter commented:
[TODO](https://github.com/maherm/steamgifts_scripts/pull/3#issuecomment-291911112) UPDATES
 1 Sync all occurs when a user hasn't visited the website via the computer for 1hr or more.
 5 Done. This is actually pretty hard to test and confirm. Probably doesn't work, but even if it doesn't it's not a gaping issue.
 6 Removed fixDatabase, unnecessary console.logs and commented code. Also adjusted crappy tampermonkey indentation.
 7 Fixed glaring bug where if you entered a GA after u bookmarked it on the same page, it wouldn't track the states (since I didn't put a click handler on the button then).

All done! The only issues now is maybe tweaking the 1hr (lower to 30min or higher to 1hr30 ?) but apart from that, it's good to go. I'm pretty satisfied with what's in there now.

maherm commented:
Great work! I'll try to have a look at it and merge it this weekend.

crazoter commented:
There's currently a display bug involving the GA data: if the GA's start timing is delayed (i.e. starting an hour from now for example), this timing is erroneously taken as the end timing (i.e. it will show ending in an hour from now). The main problem with this bug is that once the GA starts, it is shown as ended in the bookmarks.

I'm not sure if this issue is due to SG API, or due to mishandling on this userscript's side. Most likely due to SG API.

maherm commented:
Thats a known bug in SgApi (maherm/sgapi#1). I haven't found the time to fix that yet.

crazoter commented:
TIL you can exit entered GAs at https://www.steamgifts.com/giveaways/entered. Syncing the bookmarks with these buttons is currently a KIV for me.

maherm commented:
I have finally gotten to merge your changes into [my develop branch](/maherm/steamgifts_scripts/tree/develop) and ran into some issues:

1. If you delete a GA mid-train, the train is not rendered correctly afterwards. The following GA will be displayed as a new train start. F5 solves this problem

![before](https://cloud.githubusercontent.com/assets/5812806/25315521/7193154c-2856-11e7-9e9c-a15d90aef7e0.PNG)
![after](https://cloud.githubusercontent.com/assets/5812806/25315520/7192f918-2856-11e7-9e5d-2fd45a71c7cc.PNG)

2. I thought the sync feature should ensure the bookmark list is consistent over all tabs, but it doesn't. Did I get that wrong? Does it only sync the entered/owned state?

EDIT: I extended the code to reflect changes (bookmark added/removed) in other tabs by making the ```rebuildNavRows``` variable a function that checks a global ```lastChanged``` timestamp from GM_Storage and compares it to a local ```lastRebuild``` timestamp. This way, every tab always shows all bookmarks.  Could you please check whether this turns up issues with a giant bookmark list like yours?  
  
I could not figure out how to fix the first issue. Could you please try to fix that and make a new PR based on the develop branch?

I'll close this because it's merged into develop. I'm not sure whether you can reopen your own issues if I close them. So if you need this one reopened or just want to discuss anything just hit me up on steam.

crazoter commented:
1. Alright, I'll take a look at it. 

2. It hadn't crossed my mind to sync it across all tabs since I figured people would just refresh the tab or something but I guess you managed to implement it :p 

TODO: 
Fix mid-train glitch
Update syncAll (I found a new bug where if the giveaway was removed by the creator it would cause issues with the script)

crazoter commented:
Sorry maherm I had forgotten all about it :p I'll get right to it today!

crazoter commented:
EDIT: nvm I was being stupid with Git lol
